File: trajectoryTools/LUT_1D_Cartesian.m
% 1D variable density filling cartesian k-space trajectory
% for Retrospective 
%
% LUT file for pe1_order = 3 (LUT) and retrospective  
% copy and paste the LUT into the no_views table in preclinical
%
%
% Gustav Strijkers
% Amsterdam UMC
% g.j.strijkers@amsterdamumc.nl
% July 2025
%


clearvars;
close all;
clc;



%% Define dimensions, use even numbers

targetKspaceSize = 192;                 % Size of resulting k-space                 192
trajectoryLength = 256;                 % Nr of views 2                             256
densityShape = "gauss";                 % Currently only "gauss" implemented 
sigma = 8;                              % Width of the more densily filled center   8
showPlot = true;                        % Show plot true/false
output = './output/';                   % Output directory


%% Checks

mustBePosEvenInt(targetKspaceSize);
mustBePosEvenInt(trajectoryLength);
mustBePosEvenInt(sigma);



%% Calculate target k-space

kSpaceCenter = floor(targetKspaceSize/2 + 1);
extraLines = trajectoryLength-targetKspaceSize;
k = 1:targetKspaceSize;

switch densityShape

    case "gauss"

        % Gaussian function
        d = (1/(sigma*sqrt(2*pi)))*exp(-((k-kSpaceCenter).^2)./(2*sigma.^2));

end

% Make discrete
incr = 0.9;
df = round(d*incr);
while sum(df) <= extraLines
    df = round(d*incr);
    incr = incr + 0.001;
end

% Make sure that total number of k-lines equals the trajectory length
pm = true;
while sum(df) > extraLines
    loc = find(df==1);
    if pm
        df(loc(1)) = 0;
    else
        df(loc(end)) = 0;
    end
    pm = ~pm;
end
d = 1 + df;




%% Construct the zigzag trajectory

cnt = 1;
traj = zeros(targetKspaceSize,1);

for k = 1:2:targetKspaceSize

    for c = 1:d(k)
        traj(cnt) = k;
        cnt = cnt + 1;
    end

end

for k = targetKspaceSize:-2:2

    for c = 1:d(k)
        traj(cnt) = k;
        cnt = cnt + 1;
    end

end

traj = traj - targetKspaceSize/2 - 1;




%% Simulate a random k-space filling

N = 240000;

samples = randi([1 trajectoryLength], N, 1);

filling = zeros(targetKspaceSize,1);

for cnt = 1:length(samples)

    index = traj(samples(cnt)) + targetKspaceSize/2 + 1;
    filling(index) = filling(index) + 1;

end

filling = filling / sum(filling);



%% Plot the result

if showPlot

    titleFontSize = 28;
    axisLabelFontSize = 20;
    axisFontSize = 16;
    lineWidth = 3;

    fig = figure(11);
    fig.Position(3:4) = [1200,500];
    t = tiledlayout(1,2);

    nexttile;
    hold on;
    plot(traj,'LineWidth', lineWidth);
    box on;
    title("Trajectory",'FontSize', titleFontSize);
    axis([0 trajectoryLength+1 min(traj)*1.1 max(traj)*1.1 ]);
    xlabel("Sample",'FontSize',axisLabelFontSize,'FontName','Arial');
    ylabel("K-line",'FontSize',axisLabelFontSize,'FontName','Arial');
    ax = gca;
    set(ax,'FontSize',axisFontSize,'FontName','Arial')
    hold off;

    nexttile;
    hold on;
    plot(filling,'LineWidth', lineWidth);
    box on;
    title("Estimated filling of k-space",'FontSize', titleFontSize);
    axis([0 targetKspaceSize+1 0 max(filling)*1.1 ]);
    xlabel("K-line",'FontSize',axisLabelFontSize,'FontName','Arial');
    ylabel("Filling density",'FontSize',axisLabelFontSize,'FontName','Arial');
    ax = gca;
    set(ax,'FontSize',axisFontSize,'FontName','Arial')
    hold off;

end


%% Export the trajectory


filename = strcat(output,filesep,'LUT_Cartesian_1D_',num2str(targetKspaceSize),'_',num2str(trajectoryLength),'_',num2str(sigma),'.txt');
fileID = fopen(filename,'w');

for i = 1:length(traj)

    fprintf(fileID,num2str(traj(i)));
    fprintf(fileID,',\n');

end

fclose(fileID);




%% Summary of k-space filling

centerWidth = 0.2 * targetKspaceSize;  % central 20%
centerIdx = round(targetKspaceSize/2 - centerWidth/2 + 1) : round(targetKspaceSize/2 + centerWidth/2);
edgeIdx = [1:round(0.1*targetKspaceSize), round(0.9*targetKspaceSize):targetKspaceSize];

fprintf('\n--- K-space filling summary ---\n');
fprintf('Target k-space size    : %d\n', targetKspaceSize);
fprintf('Trajectory length      : %d\n', trajectoryLength);
fprintf('Density shape          : %s\n', densityShape);
fprintf('Sigma                  : %d\n', sigma);
fprintf('Mean filling           : %.4f\n', mean(filling));
fprintf('Min filling            : %.4f\n', min(filling));
fprintf('Max filling            : %.4f\n', max(filling));
fprintf('Std of filling         : %.4f\n', std(filling));
fprintf('Center/Edge fill ratio : %.2f\n', mean(filling(centerIdx)) / mean(filling(edgeIdx)));
fprintf('-------------------------------\n\n');





%% Helper function

function mustBePosEvenInt(x)

    % Throw an error unless x is:
    %   – a positive number
    %   – an integer (no fractional part)
    %   – even

    assert( isnumeric(x)             && ... % numeric input
            all(x > 0)               && ... % positive
            all(mod(x,1) == 0)       && ... % integer
            all(mod(x,2) == 0), ...         % even
            'Input must be a positive, even integer.');

end

File: trajectoryTools/LUT_1D_Cartesian.py
# 1D variable density filling cartesian k-space trajectory
# for retrospective
# 
# LUT file for pe1_order = 3 (LUT) and retrospective  
# copy and paste the LUT into the no_views table in preclinical
#
#
# Gustav Strijkers
# Amsterdam UMC
# g.j.strijkers@amsterdamumc.nl
# July 2025
#


import numpy as np
import matplotlib.pyplot as plt
import os

# Parameters

targetKspaceSize = 192         # Size of resulting k-space                 192
trajectoryLength = 256         # Nr of views 2                             256
densityShape = "gauss"         # Currently only "gauss" implemented
sigma = 8                      # Width of the more densily filled center   8
showPlot = False               # Show the plot True / False
output = './output/'           # Output folder

# Checks

def mustBePosEvenInt(x):
    assert isinstance(x, int) and x > 0 and x % 2 == 0, "Input must be a positive, even integer."

mustBePosEvenInt(targetKspaceSize)
mustBePosEvenInt(trajectoryLength)
mustBePosEvenInt(sigma)

# Calculate target k-space

kSpaceCenter = targetKspaceSize // 2
extraLines = trajectoryLength - targetKspaceSize
k = np.arange(1, targetKspaceSize + 1)

if densityShape == "gauss":
    d = (1 / (sigma * np.sqrt(2 * np.pi))) * np.exp(-((k - (kSpaceCenter + 1)) ** 2) / (2 * sigma ** 2))
else:
    raise ValueError("Unsupported density shape")

# Make discrete
incr = 0.9
df = np.round(d * incr).astype(int)
while np.sum(df) <= extraLines:
    incr += 0.001
    df = np.round(d * incr).astype(int)

# Trim to match trajectory length
pm = True
while np.sum(df) > extraLines:
    loc = np.where(df == 1)[0]
    if len(loc) == 0:
        break
    if pm:
        df[loc[0]] = 0
    else:
        df[loc[-1]] = 0
    pm = not pm

d = 1 + df


# Construct the zigzag trajectory
traj = []
for i in range(0, targetKspaceSize, 2):
    traj.extend([i + 1] * d[i])

for i in range(targetKspaceSize - 1, 0, -2):
    traj.extend([i + 1] * d[i])

traj = np.array(traj) - targetKspaceSize // 2 - 1


# Simulate random k-space filling
N = 240000
samples = np.random.randint(0, trajectoryLength, N)
filling = np.zeros(targetKspaceSize, dtype=int)

for s in samples:
    index = traj[s] + targetKspaceSize // 2
    if 0 <= index < targetKspaceSize:
        filling[index] += 1

filling = filling / np.sum(filling)


# Plotting
if showPlot:
    titleFontSize = 18
    axisLabelFontSize = 14
    axisFontSize = 12
    lineWidth = 2

    fig, axs = plt.subplots(1, 2, figsize=(12, 5))
    axs[0].plot(traj, linewidth=lineWidth)
    axs[0].set_title("Trajectory", fontsize=titleFontSize)
    axs[0].set_xlabel("Sample", fontsize=axisLabelFontSize)
    axs[0].set_ylabel("K-line", fontsize=axisLabelFontSize)
    axs[0].tick_params(labelsize=axisFontSize)
    axs[0].grid(True)

    axs[1].plot(filling, linewidth=lineWidth)
    axs[1].set_title("Estimated filling of k-space", fontsize=titleFontSize)
    axs[1].set_xlabel("K-line", fontsize=axisLabelFontSize)
    axs[1].set_ylabel("Filling density", fontsize=axisLabelFontSize)
    axs[1].tick_params(labelsize=axisFontSize)
    axs[1].grid(True)

    plt.tight_layout()
    plt.show()


# Export trajectory to text file
filename = os.path.join(output, f"LUT_cartesian_1D_{targetKspaceSize}_{trajectoryLength}_{sigma}.txt")
with open(filename, 'w') as f:
    for t in traj:
        f.write(f"{t},\n")


# Summary
centerWidth = int(0.2 * targetKspaceSize)
centerStart = targetKspaceSize // 2 - centerWidth // 2
centerEnd = centerStart + centerWidth
centerIdx = np.arange(centerStart, centerEnd)

edgeIdx = np.concatenate((np.arange(0, int(0.1 * targetKspaceSize)), 
                          np.arange(int(0.9 * targetKspaceSize), targetKspaceSize)))

ky_min = int(traj.min())
ky_max = int(traj.max())

print("\n--- K-space filling summary ---")
print(f"Target k-space size    : {targetKspaceSize}")
print(f"Trajectory length      : {trajectoryLength}")
print(f"Density shape          : {densityShape}")
print(f"Sigma                  : {sigma}")
print(f"Mean filling           : {np.mean(filling):.4f}")
print(f"Min filling            : {np.min(filling):.4f}")
print(f"Max filling            : {np.max(filling):.4f}")
print(f"Std of filling         : {np.std(filling):.4f}")
print(f"Center/Edge fill ratio : {np.mean(filling[centerIdx]) / np.mean(filling[edgeIdx]):.2f}")
print(f"ky range               : {ky_min} to {ky_max}")
print("-------------------------------\n")
